feat: AEP00004: Refactoring of MongoDB module (#48)
aneoconsulting/ArmoniK.Community: AEP/aep-00004.md

# AEP 4: Refactoring of MongoDB module : possibility to deploy a sharded database and use of Helm charts
|                   |ArmoniK Enhancement Proposal|
---:                |:---
**AEP**             | 4
**Title**           | AEP 4: Shard the MongoDB database and deploy it with a Helm chart
**Author**          | Trystan Schneider <<tschneider@aneo.fr>>
**Status**          | Draft
**Type**            | Standard
**Creation Date**   | 2024-06-26

# Abstract
This AEP proposes a new deployment for the MongoDB database. The insight is to shard (horizontal partitioning) MongoDB and deploy it with a Helm chart (the package manager for Kubernetes).

# Motivation
The motivation of this AEP is to proactively treat a previsible bottleneck on the database. 

> Currently in ArmoniK, MongoDB is deployed with a simple replica set architecture, meaning that the database is composed of multiple identical instances. However, only one instance of the database supports write operations (the primary node), resulting in a potential bottleneck on this node. Additionnaly replicating the whole database is not very efficient for read operations, as the DB engine has to parse bigger datasets.

Reviewer: ```suggestion
Currently in ArmoniK, MongoDB is deployed using a simple replica set architecture, meaning that the database is composed of multiple identical instances. However, only one instance of the database supports write operations (the primary node), resulting in a potential bottleneck on this node. Additionnaly replicating the whole database is not very efficient for read operations, as the DB engine has to parse bigger datasets.
```

> Finally, the current MongoDB module in our infrastructure's code is defined kind of manually with multiple resources being defined, which is not  necessarily the best choice as deploying a database, especially a sharded one, is a complex task. 

Reviewer: I am not sure to understant what your are writing here

Author: Currently, the way the database is deployed entirely replicates data with multiple identical replicas. And among these replicas, only one supports write operations (inserting data into the database). This where the concern of a potential bottleneck lies. Also, I wanted to mention that a non-sharded database could raise a concern on read operations too.

For the second paragraph I was highlighting that our current MongoDB deployment, which doesn't use a Helm chart, is not using the best practices, as we handle the complexity of creating all the resources needed for a MongoDB server ourselves, instead of delegating it to a Helm chart.

This sharded (or not) MongoDB deployment would be defined in our infrastructure code as a Helm release resource (called by Terraform). A Helm release is an instance of a Helm chart. Helm charts are templates that allow to deploy resources within a Kubernetes cluster in configurable way.
Using Helm charts conveys the complexity of defining a deployment towards the authors of a Helm chart who had more time to think about how MongoDB should be deployed within a Kubernetes cluster.

Instead of  manually defining a MongoDB deployment with Terraform modules as it is currently done with ArmoniK's infrastructure, Helm charts do it for us. All we have to do is to define the right configuration for the MongoDB deployment.

During our reflexion, some other alternatives to responding this issue were considered :

> - First, adding some resources and modifying some others in our Terraform module. But according to the KISS and "don't reinvent the wheel" principles, we found this solution unrelevant as others developers are able to provide us out-of-the-box MongoDB deployments, which is often a better way of building softwares as the complexity is shifted to the Helm chart definiton.

Reviewer: ```suggestion
- Adding some resources and modifying some others in our Terraform module. To respect good practices (KISS and "don't reinvent the wheel" principles), we found this solution unrelevant as others developers are able to provide out-of-the-box MongoDB deployments, which is often a better way of building softwares as the complexity is shifted to the Helm chart definiton.
```

> - Using Kubernetes operator for MongoDB, but as far as we have explored this solution, we were not able to find an operator that supports sharding.

Reviewer: ```suggestion
- Using Kubernetes operator for MongoDB. Unfortunately, no operator supporting the sharding have been found.
```

> To conclude, using a MongoDB Helm chart would allow ArmoniK to benefit from an out-of-the-box MongoDB deployment to separate the complexity of it from the sheer infrastructure code defined in Terraform, as well as benefit from the maintenance of the Helm chart used, again in the scope of dividing complexity.

Reviewer: ```suggestion
To conclude, using a MongoDB Helm chart would allow ArmoniK to benefit from an out-of-the-box MongoDB deployment to separate the complexity of it from the sheer infrastructure code defined in Terraform, as well as benefit from the maintenance of the Helm chart provided by the MongoDB community itself.
```

# Specification

## Overview

> As ArmoniK's infrastructure is described via Terraform, instead of defining multiple Kubernetes and other resources as it is currently done, we would (almost) only define a `helm_release` resource provided by the Helm provider in Terraform. The "almost" is there because we still need to define two kubernetes secrets that store the MongoDB deployment's information and users' credentials to ensure retrocompatibility with the way ArmoniK currently retrieves these data.

Reviewer: ```suggestion
Until now ArmoniK's infrastructure was mainly described using Terraform files and bash scripts. Using Helm charts 
for ArmoniK's deployement on Kubernetes will reduce the maintenance and improve the reusability of our example infrastructure by other organisation.
```

Author: ArmoniK's example infrastructure will still be described using Terraform files ! That said, this paragraph should rather be about ArmoniK's table storage than about its whole infrastructure.

> By default, ArmoniK's infrastructure would use the bitnami's Helm chart for MongoDB (sharded or not).

Reviewer: ```suggestion
In the example infrastrucutre, ArmoniK's infrastructure is using bitnami's Helm chart for MongoDB (sharded or not).
```

Here is what the Helm chart would handle indepedently with the chart used (sharded or not) :
- The creation of the Kubernetes resources needed to deploy MongoDB, including the persistent volumes and the underlying logic to use them (see [PersistentVolumeClaims documentation](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims))
- MongoDB initialization and launching

## WARNING
**As the developments and tests supporting this AEP were done using Bitnami's Helm charts, the following specifications apply to these charts and it is absolutely not guaranted that they are valid with other charts**

> Thus, we strongly advice you to use Bitnami's Helm charts for MongoDB ([sharded](https://artifacthub.io/packages/helm/bitnami/mongodb-sharded), [unsharded](https://artifacthub.io/packages/helm/bitnami/mongodb)) as well as Bitnami's MongoDB images. Note that each chart has a specific image, an image for the sharded chart and another for the unsharded chart.

Reviewer: ```suggestion
Thus, we strongly advise you to use Bitnami's Helm charts for MongoDB ([sharded](https://artifacthub.io/packages/helm/bitnami/mongodb-sharded), [unsharded](https://artifacthub.io/packages/helm/bitnami/mongodb)) as well as Bitnami's MongoDB images. Note that each chart has a specific image, an image for the sharded chart and another for the unsharded chart.
```

## Technical Specifications
**What would globally change**:
- The possibility to choose whether to deploy MongoDB sharded or not. 
- The data pods would be deployed within a StatefulSet object in Kubernetes, which is more suited for stateful applications (for more information, check the [StatefulSet documentation](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/)).

> - Since Helm takes over the whole MongoDB initialization, the configmap.tf file inside the MongoDB module, which defines the local file resources containing the initialization scripts, would be removed.

Author: If there's a need to have init scripts for MongoDB, previous versions of ArmoniK.Infra still have a Terraform module with an initialization script in it !

- As a result, the connection scheme to the service is starting with a `mongodb+srv://` scheme, which is able to perform dynamic DNS recognition of the pods (for more information, check out the [connection string documentation](https://www.mongodb.com/docs/manual/reference/connection-string/#srv-connection-format) in MongoDB).
- Since the server is now preferably accessed with SRV strings, the port output is still present for retrocompatibility concerns but hardcoded with a value of 27017 (MongoDB default port).

**What would be specific to the classic unsharded MongoDB Helm chart ?**
- It would be possible to deploy an out-of-the-box sharded MongoDB instance, with a configurable, but rather static, number of shards. We are currently working on a way to implement autoscaling.

> - TLS seems to be no longer supported.

Author: My bad, there's an error in the section's subtitle

- For the moment, it would not be possible to define custom MongoDB users only using Bitnami's MongoDB sharded chart. Only root is defined by default. We are currently working on this issue. A first solution would be to define a Kubernetes job that creates MongoDB users after the deployment is ready.

## What is database sharding and how does it work ?

> Sharding a database consists of breaking a database into several smaller independant parts, as the name implies. It is used on distributed architectures to enhance the efficiency of database operations and allieviate tension on nodes.

Author: A sharded database infrastructure is complex to set up and maintain and consumes more resources. There can be cases where sharding is unappropriate (think about Montparnasse !)

Bitnami's mongodb-sharded Helm chart offers an out-of-the box infrastructure setup needed to host a sharded MongoDB instance and mainly creates the following Kubernetes objects :

- For the shards, each of them is deployed through a StatefulSet. Each StatefulSet contains the replicas associated with each shard. For example, if you want to deploy 3 shards, each being replicated once, you would have in Kubernetes, 3 StatefulSet's : `shard-0, shard-1, shard-2` each of them containing 2 pods, `data-0` and `data-1`.
- The config servers are also deployed through a unique StatefulSet managing all the config servers.
- The routers are deployed through a unique Deployment, since they are stateless proccesses, managing all the mongos instances.

Finally, two Services are created :
- A headless Service, which is used for the internal functionning of the sharded cluster in Kubernetes. For example, to perform a write operation, the routers need to know how to access to a particular shard pod. (for more information on headless Services, see the [Kubernetes documentation](https://kubernetes.io/docs/concepts/services-networking/service/#headless-services))

> - A "classical" Service, which is used by MongoDB clients to access a mongos instance.

Author: It will depend on the configuration you choose. I will add that the number of config svr replicas and mongos to deploy is also configurable.

### Back to MongoDB
Once these components are deployed inside a Kubernetes Cluster, a database must be created and sharding enabled on it. To perform that, a connection to a router instance must be established :

Assuming the Service to access the routers is called "mongodb-sharded" and listening on MongoDB default port:
```bash
mongosh -u $MONGODB_ADMIN_USER -p $MONGODB_ADMIN_PASSWORD mongodb://mongodb-sharded:27017
```
Then to create a database and enable sharding on it :
```javascript

> sh.enableSharding("our_database")

Reviewer: where do you need to use this command?

Author: This the standard command used in mongosh. It might be the responsibility of ArmoniK users to enable sharding on their database. @lemaitre-aneo  do you agree ?

```

> Note that these actions are preferably performed by a program.

Reviewer: ```suggestion
Note is it better to automatize these actions.
```

> We have now a database "our_database" supporting sharding on collection-level. It's nice but it's not the point of sharding a database. To have a really nice sharding, we need to define a shard key.

Reviewer: not in the aep but the last part which should be defined before. It is a new concept.

### Shard Keys

> Shard keys are basically indexes (simple or compounded) that will be used to determine the way data is distributed among the shards.

Reviewer: ```suggestion
Shard keys are basically indexes (simple or compounded) that will be used to determine the way data are distributed among the shards.
```

Defining a relevant shard key is crucial when sharding a database. A bad shard key could result in a bottleneck and even worsen the efficiency of your database. For more information on how to choose a shard key check the [Choose a Shard Key documentation](https://www.mongodb.com/docs/manual/core/sharding-choose-a-shard-key/). The shard key choice falls to the database administrator.

MongoDB proposes two types of shard keys :

> - Hashed keys, which are better at evenly distribute data among shards but worse for multiple-documents read operations, as they likely distribute documents with close shard key value on different shards. (see [Hashed Sharding documentation](https://www.mongodb.com/docs/manual/core/hashed-sharding/))

Reviewer: ```suggestion
- Hashed keys, which are better to evenly distribute data among shards but increase the read operations for multiple-documents, as they likely distribute documents with close shard key value on different shards. (see [Hashed Sharding documentation](https://www.mongodb.com/docs/manual/core/hashed-sharding/))
```

> - Range keys, which are not ideal for evenly distribute data among shards but better for multiple-documents read operations. (see [Ranged Sharding documentation](https://www.mongodb.com/docs/manual/core/ranged-sharding/))

Reviewer: ```suggestion
- Range keys, which are not ideal for evenly distribute data among shards but more adapted for multiple-documents read operations. (see [Ranged Sharding documentation](https://www.mongodb.com/docs/manual/core/ranged-sharding/))
```

> 

Reviewer: So which one is more adpated for Armonik?

### Shard the collections
That being said, you can now shard the collections you want.
Let's say that inside the "our_database" database, you want to shard the collection named "collection_1", with a hashed shard key on the "_id" field:
```javascript

> sh.shardCollection("our_database.collection_1", {_id: "hashed})

Reviewer: ```suggestion
sh.shardCollection("our_database.collection_1", {_id: "hashed"})
```

```

For more information on sharding possibility, check out the [sh.shardCollection methode documentation](https://www.mongodb.com/docs/manual/reference/method/sh.shardCollection/#mongodb-method-sh.shardCollection)

> ### Reference

Reviewer: ```suggestion
### References
```

For more information on how sharding works in MongoDB, check out their [official sharding documentation](https://www.mongodb.com/docs/manual/sharding/)

# Backwards Compatibility

## With ArmoniK Infrastructure (IMPORTANT)

> Currently, the main concern on backwards compatibility is about data persistency in MongoDB, as we haven't focused on the possibility to pass down custom persistent volumes to the Helm release resource. 

Reviewer: ```suggestion
Currently, the main concern on backwards incompatibility is about data persistency in MongoDB, as we haven't focused on the possibility to pass down custom persistent volumes to the Helm release resource. 
```

> Which means that if your database contains data, if you want to upgrade the infrastructure version to deploy MongoDB (sharded or not doesn't matter) with a Helm chart, one of these two situations will :

Reviewer: ```suggestion
Which means that if your database contains data, if you want to upgrade the infrastructure version to deploy MongoDB (sharded or not doesn't matter) with a Helm chart, one of these two situations will arise:
```

- If you are using Persistent Volumes with a reclaim policy set to "Retain", you will be able to keep your data but unable to automatically provide it to the new MongoDB pods freshly created by the Helm release.
- Your data will be lost if you are using Persistent Volumes with a reclaim policy set to "Delete", which is the default value with dynamic provisioning.

## With ArmoniK
This refactoring would not be backward-compatible with previous versions of ArmoniK since we would prefer to not set default values for the Helm chart version. Which means that calls to this refactored MongoDB module must define the `helm_chart_version` parameter, which is not implemented in previous versions of ArmoniK.

## With ArmoniK Core
This refactoring would be backward-compatible with previous versions of ArmoniK Core but would not be optimal in case of a sharded MongoDB instance, since these previous versions do not implement the database commands to enable sharding. 

# Security Implications

> The unsharded Bitnami's MongoDB Helm chart does not generate TLS certificate for clients. The sharded one does not seem to support TLS at all.

Reviewer: This is a very big problem but here we should probably mention that our deployement are only example and could be/should be different in a production environment  and respect the security policies of each company.

Author: You are talking about the sharded Helm chart not supporting TLS ?

# AEP 4: Refactoring of MongoDB module : possibility to deploy a sharded database and use of Helm charts
|                   |ArmoniK Enhancement Proposal|
---:                |:---
**AEP**             | 4
**Title**           | AEP 4: Sharded MongoDB database deployment with the help of a Helm chart
**Author**          | Trystan Schneider <<tschneider@aneo.fr>>
**Status**          | Draft
**Type**            | Standard
**Creation Date**   | 2024-06-26

# Abstract
This AEP proposes a new deployment for the MongoDB database. The insight is to shard (horizontal partitioning) MongoDB and deploy it with a Helm chart (the package manager for Kubernetes).

# Motivation
The motivation of this AEP is to proactively treat a previsible bottleneck on the database. 

> Currently in ArmoniK, MongoDB is deployed using a simple replica set architecture, meaning that the database is composed of multiple identical instances. However, only one instance of the database supports write operations (the primary node), resulting in a potential bottleneck on this node. Additionnaly replicating the whole database is not very efficient for read operations, as the DB engine has to parse bigger datasets.

Reviewer: la derniere phrase, je ne suis pas sur de la comprendre. Si tu ne mets pas de replica mais que tu n'as qu'une seule DB le probleme sera le meme. Le vrai probleme c'est que l'infra actuelle meme en replica n'utilise que le "primary node" et que les replicats ne sont là que pour la persistance des données.

> Finally, the current MongoDB module in our infrastructure's code is defined manually, which is not  necessarily the best choice as deploying a database, especially a sharded one, is a complex task. 

Reviewer: ```suggestion
Finally, the current MongoDB module in our infrastructure's code is defined manually, which is not  necessarily the best choice, deploying a database is a complex task, especially a sharded one. 
```

# What is database sharding and how does it work ?

> Sharding a database consists of breaking a database into several smaller independant parts, as the name implies. It is used on distributed architectures to enhance the efficiency of database operations and allieviate tension on nodes.

Reviewer: ```suggestion
Sharding a database consists of splitting a database into several smaller independant parts. It is mostly used on distributed architectures to enhance the efficiency of database operations and allieviate tension on nodes.
```

## Sharding with MongoDB
MongoDB natively offers the possibility to set up a sharded database.

> In MongoDB, here are the basics of how sharding is implemented :

Reviewer: ```suggestion
In MongoDB, the basics to implement the sharding are:
```

Firstly, we need to distinguish three main components in a sharded MongoDB infrastructure. Each of these components is actually an independant process and can have multiple instances in a sharded MongoDB cluster :
- The shards, which are the instances that contain the data (see [Shard documentation](https://www.mongodb.com/docs/manual/core/sharded-cluster-shards/))

> - The routers (also called mongos), which are the instances that route database operations to the good shard(s). They serve as interface between database clients and the shards. (see [Routers (mongos) documentation](https://www.mongodb.com/docs/manual/core/sharded-cluster-query-router/))

Reviewer: ```suggestion
- The routers (also called mongos), which are the instances that route database operations to the proper shard(s). They are serving as interface between the database clients and the shards. (see [Routers (mongos) documentation](https://www.mongodb.com/docs/manual/core/sharded-cluster-query-router/))
```

> - The config servers, which store the sharded cluster's metadata : how data is distributed, as well as authentication information and RBAC settings. They also manage distributed locks. They are used by the routers to determine how to route queries. 

Reviewer: ```suggestion
- The config servers, which store the sharded cluster's metadata: how data are distributed, as well as authentication information and RBAC settings. They also manage distributed locks. They are used by the routers to determine how to route queries. 
```

All these components can be deployed through a MongoDB replicaset, thanks to the replication mechanisms provided by MongoDB. 

To better envision how a sharded architecture changes from a replica set architecture, let's take a look at the schemas below.

## Simple standalone architecture
![standalone schema](./images/standalone.png)

> In a standalone architecture, only one MongoDB instance (mongod process) is available for all operations. The database user directly reaches that instance. If this instance somehow becomes unavailable data cannot be accessed.

Reviewer: ```suggestion
In a standalone architecture, only one MongoDB instance (mongod process) is available for all operations. The database user reache directly that instance. If that instance becomes unavailable data cannot be accessed.
```

## Single replica set architecture 
![standalone schema](./images/replicaset.png)

> A replica set is a set of several MongoDB instances (mongod processes) that contain the same data. A MongoDB instance that is part of a replica set is called a replica set member. In the schema above, the replica set has 3 members : one primary and two secondary. Only the primary member support write operations and are responsible of propagating data to the other members. Primary member supports read operations too. Secondary members only support read operations. A replica set cannot have more than one primary node. 

Reviewer: ```suggestion
A replica set is a set of several MongoDB instances (mongod processes) that contain the same data. A MongoDB instance that is part of a replica set is called a replica set member. In the schema above, the replica set has 3 members: one primary and two secondary. Only the primary member support write operations and are responsible to propagate data to the other members. Primary member also supports read operations. Secondary members support only read operations. A replica set cannot have more than one primary node. 
```

> What is interesting to see is that in a replica set architecture, the database user still has to reach replica set members (MongoDB instances) directly. depending on the type of operation wanted.

Reviewer: ```suggestion
What is interesting to see is that in a replica set architecture, the database user can always access replica set members (MongoDB instances) in reading mode.
```

If one or two members becomes unavailable, data are still accessible. But if the primary node becomes unavailable, it is not possible to perform write operations in the database.

## Sharded architecture
![sharding schema](./images/sharding.png)

What we can see is that a sharded database consists of several replica sets each containing only parts of the data. These replica sets are agnostic of each other. Also, from the database user point of view a sharded database is transparent, as if it was communicating with a simple standalone (or replica set if the user wants to specify a read preferency) since the user only communicates with the router.

> Hence, the real database is actually the concatenation of numerous and independant logical databases (three in our example) in the MongoDB cluster. The router instance is reponsible of this concatenation work when the user requests multi-documents read operations. 

Reviewer: ```suggestion
Hence, the real database is actually the concatenation of numerous and independant logical databases (three in our example) in the MongoDB cluster. The router instance is reponsible for the concatenation work needed when the user requests multi-documents located on multiple shards. 
```

> The router is also able to perform read preferences when querrying a specific shard. For readability concerns, the arrows representing distributed read operations point at the replica sets but keep in mind that the router is able to query the all the members of a shard individually. 

Reviewer: ```suggestion
The router is also able to perform read preferences when querrying a specific shard. For readability concerns, the arrows representing distributed read operations point at the replica sets but keep in mind that the router is able to query all the members of a shard individually. 
```

> In a sharded cluster, the mongos instance kind of endorses the role of the database user in the previous architectures shown.

Reviewer: ```suggestion
In a sharded cluster, the mongos instanceis similar to the role of the database user presented in the previous architectures.
```

Note that the config server is also deployed as a replica set. It is not clear in the schema above because we didn't want to overload it with too much information, as we are focusing on sharding user-intended database.

> Finally, also note that multiple mongos routers can be deployed in the cluster too but no need for replica set as they are stateless.

Reviewer: ```suggestion
It is important to notice that multiple mongos routers can also be deployed in the cluster but there are no need for replica set as they are stateless.
```

# Rationale
Sharding the database would alleviate the tension on all the nodes. For the primary nodes, the write operations would be distributed among other shards' primary node, and for the secondary nodes there can be some situations where only the right shard is queried (it will be explicited below in this AEP) and as mentionned above, the read operations are faster on a sharded cluster since the data is distributed among multiple instances.

> This sharded (or not) MongoDB deployment would be defined in example our infrastructure code as a Helm release resource (called by Terraform). A Helm release is an instance of a Helm chart. Helm charts are templates that allow to deploy resources within a Kubernetes cluster in configurable way.

Reviewer: ```suggestion
This sharded (or not) MongoDB deployment would be defined in an example of our infrastructure code as a Helm release resource (called by Terraform). A Helm release is an instance of a Helm chart. Helm charts are templates that allow to deploy resources within a Kubernetes cluster in configurable way.
```

> Using Helm charts conveys the complexity of defining a deployment towards the authors of a Helm chart, in accordance with the "Don't reinvent the wheel" principle.

Reviewer: ```suggestion
Using Helm charts delegate the complexity of defining a deployment towards the authors of the Helm chart, in accordance with the "Don't reinvent the wheel" principle.
```

> Instead of  manually defining all the Kubernetes resources and their configuration for a MongoDB deployment within Terraform modules as it is currently done with ArmoniK's example infrastructure, Helm charts allow us to focus only on passing the good configuration for our MongoDB deployment, which is already not a simple task. Nevertheless, it is possible that some Kubernetes resources or bash scripts will still need to be manually defined inside Terraform for configuration purposes. Helm charts can offer that possibility.

Reviewer: ```suggestion
Instead of manually defining all the Kubernetes resources and their configuration for a MongoDB deployment within Terraform modules as it is currently done with ArmoniK's example infrastructure, Helm charts allow us to focus only on passing the desired configuration for our MongoDB deployment. Nevertheless, it is possible that some Kubernetes resources or bash scripts will still be needed inside Terraform for specific configuration purposes. Helm charts is also offering that capacity.
```

> Some other alternatives to responding this issue were considered :

Reviewer: ```suggestion
Some other alternatives to answer this problematic were considered :
```

- Adding some resources and modifying some others in our Terraform module. To respect good practices (KISS and "don't reinvent the wheel" principles), we found this solution unrelevant as others developers are able to provide us out-of-the-box MongoDB deployments, which is often a better way of building softwares as the complexity is shifted to the Helm chart definiton.
- Using Kubernetes operator for MongoDB. Unfortunately, no operator supporting sharding have been found.

To conclude, using a MongoDB Helm chart would allow ArmoniK to benefit from an out-of-the-box MongoDB deployment to separate the complexity of it from the sheer infrastructure code defined in Terraform, as well as benefit from the maintenance of the Helm chart provided by the MongoDB community itself.

# Specification
## Overview
Until now ArmoniK's table storage was described with multiple Kubernetes resources and bash scripts. Using Helm charts for table storage will reduce the maintenance and improve the reusability of our example infrastructure by other organisations.

> In the example infrastructure, Bitnami's Helm chart for MongoDB (sharded or not) will be used.

Reviewer: mettre un lien pour le projet bitnami

Here is what the Helm chart would handle indepedently with the chart used (sharded or not) :
- The creation of the Kubernetes resources needed to deploy MongoDB, including the persistent volumes and the underlying logic to use them (see [PersistentVolumeClaims documentation](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims))
- MongoDB initialization and launching

## WARNING

> **As the developments and tests supporting this AEP were done using Bitnami's Helm charts, the following specifications apply to these charts and it is absolutely not guaranted that they are valid with other charts**

Reviewer: ```suggestion
**As the developments and tests supporting this AEP were done using Bitnami's Helm charts, the following specifications apply to these charts and it is absolutely not guaranted that they are valid with other charts or images**
```

> Thus, we strongly advise you to use Bitnami's Helm charts for MongoDB ([sharded](https://artifacthub.io/packages/helm/bitnami/mongodb-sharded), [unsharded](https://artifacthub.io/packages/helm/bitnami/mongodb)) as well as Bitnami's MongoDB images. Note that each chart has a specific image, an image for the sharded chart and another for the unsharded chart.

Reviewer: je ne suis pas totalement en accord avec cette phrase. En effet, il n'y a pas de verification visible des images avec docker scout. Conseiller fortement leur usage est donc un peu ... dangereux.

- The deployment would be exposed with a headless Kubernetes service, which is more suited for stateful applications that don't require load-balancing, as it allows to reach the pods directly as they have a persistent and identifying DNS record within the Kubernetes cluster (see [headless services documentation](https://kubernetes.io/docs/concepts/services-networking/service/#headless-services) in Kubernetes).
- As a result, the connection scheme to the service is starting with a `mongodb+srv://` scheme, which is able to perform dynamic DNS recognition of the pods (for more information, check out the [connection string documentation](https://www.mongodb.com/docs/manual/reference/connection-string/#srv-connection-format) in MongoDB).
- Since the server is now preferably accessed with SRV strings, the port output is still present for retrocompatibility concerns but hardcoded with a value of 27017 (MongoDB default port).

**What would be specific to the sharded MongoDB Helm chart ?**

> - It would be possible to deploy an out-of-the-box sharded MongoDB instance, with a configurable, but rather static, number of shards. We are currently working on a way to implement autoscaling. The number of config servers replicas and mongos "replicas" would also be configurable.

Author: Je pense que l'autoscaling (horizontal) est une conséquence plus que logique au sharding de base de données, donc le mentionner me parait pertinent

- TLS seems to be no longer supported.
- The *database* namespace in MongoDB as well as non-root MongoDB users are created via a custom initialization script. 

## Inside Kubernetes

> In our case, since the MongoDB cluster is deployed within a Kubernetes cluster thanks to a Helm chart, we need to describe the infrastructure Helm creates to manage the different components of the MongoDB cluster.

Author: Oui bonne remarque, j'ai modifié le paragraphe.
